dev/TEAL_opcodes.md

```diff
@@ -0,0 +1,440 @@
+---
+numbersections: true
+title: "Algorand Transaction Execution Approval Language, Opcodes"
+date: \today
+abstract: >
+  Algorand allows transactions to be effectively signed by a small program. If the program evaluates to true then the transaction is allowed. This document defines the language opcodes and byte encoding.
+---
+
+# Opcodes
+
+Ops have a 'cost' of 1 unless otherwise specified.
+
+
+## err
+
+- Opcode: 0x00 
+- Pops: _None_
+- Pushes: _None_
+- Error. Panic immediately. This is primarily a fencepost against accidental zero bytes getting compiled into programs.
+
+## sha256
+
+- Opcode: 0x01 
+- Pops: *... stack*, []byte
+- Pushes: []byte
+- SHA256 hash of value, yields [32]byte
+- **Cost**: 7
+
+## keccak256
+
+- Opcode: 0x02 
+- Pops: *... stack*, []byte
+- Pushes: []byte
+- Keccac256 hash of value, yields [32]byte
+- **Cost**: 26
+
+## sha512_256
+
+- Opcode: 0x03 
+- Pops: *... stack*, []byte
+- Pushes: []byte
+- SHA512_256 hash of value, yields [32]byte
+- **Cost**: 9
+
+## ed25519verify
+
+- Opcode: 0x04 
+- Pops: *... stack*, {[]byte A}, {[]byte B}, {[]byte C}
+- Pushes: uint64
+- for (data, signature, pubkey) verify the signature of the data against the pubkey => {0 or 1}
+- **Cost**: 1900
+
+## rand
+
+- Opcode: 0x05 
+- Pops: _None_
+- Pushes: uint64
+- push random uint64 to stack
+- **Cost**: 3
+
+Random number generator based on the ChaCha20 algorithm. Seeded with the previous block's `Seed` value and the current transaction ID.
```

Comment:
Is there a rationale for seeding with the current transaction ID?

In general, we should be wary of providing this instruction--it may be theoretically impossible for the Algorand TEAL execution environment to provide applications with truly unbiased randomness.

Reply:
(Also might be good to add reference to ChaCha20 spec.)